src/avro_json_encoder.erl

```diff
@@ -123,24 +128,44 @@ enc(Lkup, Type, Union) when ?IS_UNION_TYPE(Type) ->
   [Encoded].
 
 %% @private
-optional_field(_Key, Default, Default, _MappingFun) -> [];
-optional_field(Key, Value, _Default, MappingFun) -> [{Key, MappingFun(Value)}].
+optional_field(aliases,   _, _, _, #{canon := true}) -> [];
+optional_field(default,   _, _, _, #{canon := true}) -> [];
+optional_field(doc,       _, _, _, #{canon := true}) -> [];
+optional_field(namespace, _, _, _, #{canon := true}) -> [];
+optional_field(order,     _, _, _, #{canon := true}) -> [];
+optional_field(_Key, Default, Default, _MappingFun, _Opt) -> [];
+optional_field(Key, Value, _Default, MappingFun, _Opt) ->
+    [{Key, MappingFun(Value)}].
+
+% type, name, fields, symbols, items, values, size
 
 %% @private
-do_encode_type(Name, EnclosingNamespace) when ?IS_NAME(Name) ->
+do_encode_type(Name, EnclosingNamespace, #{canon := true} = Opt)
+  when ?IS_NAME(Name) ->
+
+  {ShortName, Namespace} = case split_fullname(Name) of
```

Comment:
Not sure why you need call split_fullname/1 here, 
Can't we do it this way?
```
MaybeShortName =
  case avro:split_type_name(Name, EnclosingNamespace) of
    {ShortName, EnclosingNamespace} -> ShortName;
    {_ShortName, _AnotherNamespace} -> Name
  end,
NameToEncode = case IsCanonical of
  true -> Name;
  false -> ShortName
end,
```
`Name` is always a full-name unless there is no namespace.

Reply:
The above comment is to merge two clauses.
we can also chose to not do anything else in this clause (when `#{canon := true}`), just return`encode_string(Name);`

Reply:
Removed unnecessary code.